source/AtlasGlyphs.cpp

```diff
@@ -0,0 +1,403 @@
+	if(Screen::Width() != screenWidth || Screen::Height() != screenHeight)
+	{
+		screenWidth = Screen::Width();
+		screenHeight = Screen::Height();
+		GLfloat scale[2] = {2.f / screenWidth, -2.f / screenHeight};
+		glUniform2fv(scaleI, 1, scale);
+	}
+	
+	GLfloat textPos[2] = {
+		static_cast<float>(x - 1.),
+		static_cast<float>(y - baseline)};
+	int previous = 0;
+	bool underlineChar = false;
+	const int underscoreGlyph = Glyph('_');
+	
+	for(size_t pos = 0; pos < str.length(); pos = Font::NextCodePoint(str, pos))
+	{
+		// TODO handle composing codepoints
+		char32_t c = Font::DecodeCodePoint(str, pos);
+		if(c == '_')
+		{
+			underlineChar = Font::ShowUnderlines();
+			continue;
+		}
+		
+		int glyph = Glyph(c);
+		if(!glyph)
+		{
+			textPos[0] += space;
+			continue;
+		}
+		
+		glUniform1i(glyphI, glyph);
+		glUniform1f(aspectI, 1.f);
+		
+		textPos[0] += advance[previous * GLYPHS + glyph] + KERN;
+		glUniform2fv(positionI, 1, textPos);
+		
+		glDrawArrays(GL_TRIANGLE_STRIP, 0, 4);
+		
+		if(underlineChar)
+		{
+			glUniform1i(glyphI, underscoreGlyph);
+			glUniform1f(aspectI, static_cast<float>(advance[glyph * GLYPHS] + KERN)
+				/ (advance[underscoreGlyph * GLYPHS] + KERN));
+			
+			glUniform2fv(positionI, 1, textPos);
+			
+			glDrawArrays(GL_TRIANGLE_STRIP, 0, 4);
+			underlineChar = false;
+		}
+		
+		previous = glyph;
+	}
+	
+	glBindVertexArray(0);
+	glUseProgram(0);
+}
+
+
+
+double AtlasGlyphs::Width(const string &str) const
+{
+	int width = 0;
+	int previous = 0;
+	
+	for(size_t pos = 0; pos < str.length(); pos = Font::NextCodePoint(str, pos))
+	{
+		char32_t c = Font::DecodeCodePoint(str, pos);
+		if(c == '_')
+			continue;
+		
+		int glyph = Glyph(c);
+		if(!glyph)
+			width += space;
+		else
+		{
+			width += advance[previous * GLYPHS + glyph] + KERN;
+			previous = glyph;
+		}
+	}
+	width += advance[previous * GLYPHS];
+	
+	return width;
+}
+
+
+
+double AtlasGlyphs::LineHeight() const
+{
+	return height;
+}
+
+
+
+double AtlasGlyphs::Space() const
+{
+	return space;
+}
+
+
+
+double AtlasGlyphs::Baseline() const
+{
+	return baseline;
+}
+
+
+
+size_t AtlasGlyphs::FindUnsupported(const string &str, size_t pos) const
+{
+	if(Font::CodePointStart(str, pos) != pos)
+		pos = Font::NextCodePoint(str, pos);
+	for(; pos < str.length(); pos = Font::NextCodePoint(str, pos))
+	{
+		char32_t c = Font::DecodeCodePoint(str, pos);
+		// ASCII
+		if(c >= 32 && c < 32 + GLYPHS - 3 && c != '\'' && c != '"')
+			continue;
+		// Curly quotes
+		if(c == Font::LEFT_SINGLE_QUOTATION_MARK)
+			continue;
+		if(c == Font::RIGHT_SINGLE_QUOTATION_MARK)
+			continue;
+		if(c == Font::LEFT_DOUBLE_QUOTATION_MARK)
+			continue;
+		if(c == Font::RIGHT_DOUBLE_QUOTATION_MARK)
+			continue;
+		// Vertical rectangle
+		if(c == Font::WHITE_VERTICAL_RECTANGLE)
+			continue;
+		return pos;
+	}
+	return str.length();
+}
+
+
+
+int AtlasGlyphs::Glyph(char32_t c)
+{
+	// ASCII
+	if(c >= 32 && c < 32 + GLYPHS - 3 && c != '\'' && c != '"')
```

Comment:
Maybe add a `static const int ___ = 32` to help name this magic 32. Though anyone who is working with fonts likely knows its role without hints.

Reply:
Inherited from original code, 32 is the ASCII code for space ' '. I can clarify it.